feat: gas profile by parameter

jakmeier opened the pull request:
This resolves #8033, completing the gas profile by parameter effort.

The main achievement here is that we can now  look at gas profiles and
know exactly for which parameter it has been spent.

Code wise, the old `enum Cost` used in profiles only is finally gone.

The tricky parts in this PR are to make sure that old profiles stored on DB
are still understood and presented correctly in RPC calls.

nagisa commented on core/primitives-core/src/config.rs:
```diff
@@ -256,337 +258,114 @@ pub struct ViewConfig {
+    Debug,
+    PartialOrd,
+    Ord,
+    EnumCount,
```
This derive is now superfluous as `enum_map::Enum::LENGTH` provides the same information for this enum.

jakmeier replied:
addresssed in https://github.com/near/nearcore/pull/8115

nagisa commented on core/primitives-core/src/profile.rs:
```diff
@@ -1,88 +1,59 @@
-        for i in 0..DataArray::LEN {
-            self.data[i] = self.data[i].saturating_add(other.data[i]);
+    pub fn merge(&mut self, other: &ProfileDataV2) {
+        for (cost, gas) in self.actions_profile.iter_mut() {
+            *gas = gas.saturating_add(other.actions_profile[cost]);
         }
```
Consider

```suggestion
        for ((_, gas), (_, other_gas)) in self.actions_profile.iter_mut().zip(other.actions_profile.iter()) {
            *gas = gas.saturating_add(*other_gas);
        }
```

jakmeier replied:
Hm, feels more complicated to achieve the same thing? Anything wrong with the original code?

nagisa replied:
Indexing of the `other.actions_profile[cost]` is a panicky operation, and it needs to bound check every time too (as the indices are going upwards). The `zip` approach, avoids both of these problems by a) necessarily picking a handling mechanism (truncation) when the lengths mismatch, and b) being an iterator, thus not needing bound checks.

jakmeier replied:
Ah, these are convincing arguments. Updated!

nagisa commented on core/primitives-core/src/profile.rs:
```diff
@@ -94,37 +65,170 @@ impl ProfileData {
+/// canonical representation is not required here.
+const fn borsh_ext_index(ext: ExtCosts) -> usize {
+    match ext {
+        ExtCosts::base => 0,
```
We could keep the enum discriminants explicit and stable by specifying them in the enum declaration too, as such:

```
enum ExtCosts {
    base = 0,
    constract_loading_base = 1,
    ...
}
```

It isn’t clear to me whether this mapping approach achieves much over that other approach. But this doesn’t seem all too onerous either, so maybe it is okay to leave this as-is.

jakmeier replied:
I liked the discriminants approach and changed it.

nagisa commented on core/primitives-core/src/profile.rs:
```diff
@@ -157,146 +261,291 @@ impl fmt::Debug for ProfileData {
-    ExtCost { ext_cost_kind: ExtCosts },
-    WasmInstruction,
-}
+mod profile_v1 {
```
Any reason to make this into a module? Maybe consider moving this into a separate file?

jakmeier replied:
It is now a separate file.

nagisa commented on core/primitives/res/runtime_configs/parameters.txt:
```diff
@@ -119,6 +119,7 @@ wasm_storage_iter_next_base: 0
 wasm_storage_iter_next_key_byte: 0
 wasm_storage_iter_next_value_byte: 0
 wasm_touching_trie_node: 16_101_955_926
+wasm_read_cached_trie_node: 2_280_000_000
```
Why did this move?

jakmeier replied:
See https://github.com/near/nearcore/pull/8115

jakmeier commented:
Hey @nagisa thanks for reviewing! Sorry, I only created this draft PR to run tests on CI, it builds on the changes in https://github.com/near/nearcore/pull/8115 which have been hanging in review for 2 weeks. This probably made it a bit harder to review than necessary.

Now that the other PR is merged, I will rebase this PR, address your comments and then mark it as ready for review afterwards.

jakmeier commented:
Should be almost ready now but I want to add some tests for the RPC view output. I'll first need to work on more urgent things though...

jakmeier commented:
This should finally be ready for review.

@nagisa I think I addressed all your review comments on the draft. I would appreciate another review.

cc @akashin since this affects work on parameter weights

jakmeier commented on core/primitives/src/transaction.rs:
```diff
@@ -366,11 +366,12 @@ pub struct ExecutionOutcome {
+    /// V2: With ProfileData by legacy `Cost` enum
+    V2(ProfileDataV1),
+    // V3: With ProfileData by gas parameters
+    V3(ProfileDataV2),
```
Yes, I changed according to your suggestion. This means there is no `ProfileDataV1`, which makes sense since there is no profile data for `ExecutionMetadata::V1`.

jakmeier commented:
@nagisa do you think you will have time to look at this and are the right person to review it? I can also ask someone else but since you have already looked at an early version once (back in early December) you are the natural candidate.

nagisa commented:
Aha, thanks for the reminder, I’m looking at it now.

nagisa commented on core/primitives-core/src/profile.rs:
```diff
@@ -157,146 +202,32 @@ impl fmt::Debug for ProfileData {
-        &mut self.data[index.profile_index()]
-    }
-}
+mod profile_v2;
```
The conventional style and the one that we have been following across nearcore, I believe, is to have `mod` definitions at the top of the file, after the `use` items.

nagisa commented on core/primitives/src/views.rs:
```diff
@@ -1265,73 +1265,102 @@ impl Default for ExecutionMetadataView {
+            // Can't `sort_by_key` here because lifetime inference in
+            // closures is limited.
+            costs.sort_by(|lhs, rhs| {
+                lhs.cost_category.cmp(&rhs.cost_category).then(lhs.cost.cmp(&rhs.cost))
```
I suspect clippy would complain about the `then` here, possibly saying something about it being executed regardless of whether the first check passed or not. I suspect the compiler optimizes this out well either way, though…
